examples/model_service_template/mxnet_vision_batching.py

```diff
@@ -0,0 +1,225 @@
+        :return:
+        """
+        self._context = context
+        self._batch_size = context.system_properties["batch_size"]
+        self.initialized = True
+
+        properties = context.system_properties
+        model_dir = properties.get("model_dir")
+        gpu_id = properties.get("gpu_id")
+
+        signature_file_path = os.path.join(model_dir, "signature.json")
+        if not os.path.isfile(signature_file_path):
+            raise RuntimeError("Missing signature.json file.")
+
+        with open(signature_file_path) as f:
+            self.signature = json.load(f)
+
+        model_files_prefix = self.get_model_files_prefix(context)
+        archive_synset = os.path.join(model_dir, "synset.txt")
+        if os.path.isfile(archive_synset):
+            synset = archive_synset
+            self.labels = [line.strip() for line in open(synset).readlines()]
+
+        data_names = []
+        data_shapes = []
+        for input_data in self.signature["inputs"]:
+            data_name = input_data["data_name"]
+            data_shape = input_data["data_shape"]
+
+            # Set batch size
+            data_shape[0] = self._batch_size
+
+            # Replace 0 entry in data shape with 1 for binding executor.
+            for idx in range(len(data_shape)):
+                if data_shape[idx] == 0:
+                    data_shape[idx] = 1
+
+            data_names.append(data_name)
+            data_shapes.append((data_name, tuple(data_shape)))
+
+        checkpoint_prefix = "{}/{}".format(model_dir, model_files_prefix)
+
+        # Load MXNet module
+        self.mxnet_ctx = mx.cpu() if gpu_id is None else mx.gpu(gpu_id)
+        sym, arg_params, aux_params = mx.model.load_checkpoint(checkpoint_prefix, self.epoch)
+
+        # noinspection PyTypeChecker
+        self.mx_model = mx.mod.Module(symbol=sym, context=self.mxnet_ctx,
+                                      data_names=data_names, label_names=None)
+        self.mx_model.bind(for_training=False, data_shapes=data_shapes)
+        self.mx_model.set_params(arg_params, aux_params, allow_missing=True, allow_extra=True)
+
+    def inference(self, model_input):
+        """
+        Internal inference methods for MXNet. Run forward computation and
+        return output.
+
+        :param model_input: list of NDArray
+            Preprocessed inputs in NDArray format.
+        :return: list of NDArray
+            Inference output.
+        """
+        Batch = namedtuple('Batch', ['data'])
+        if self.error is not None:
+            return None
+
+        self.mx_model.forward(Batch([model_input]), is_train=False)
+        outputs = self.mx_model.get_outputs()
+        for d in outputs:
+            if isinstance(d, list):
+                for n in outputs:
+                    if isinstance(n, mx.ndarray.ndarray.NDArray):
+                        n.wait_to_read()
+            elif isinstance(d, mx.ndarray.ndarray.NDArray):
+                d.wait_to_read()
```

Comment:
I don't think we need wait_to_read() here.
I think ndarray.split() is imperative call, when we got res, all the read should be synced already.

Reply:
True. Any ndarray op should probably be sufficient. I can test and remove this.